### DISCUSSION TITLE
[TIP] Instructions for using Edit Config for adding Widgets

### DISCUSSION THREAD

**Discussion by poke0** [Lissy93/dashy#796] [Hot Tips]
I struggled with adding the SABnzbd last night but finally figured it out and thought I would document my process in case you wanted to share it with the community.  I'm sure you developer types have easier ways to manage your configs but us mere mortals have to do it the hard way (I'm not complaining, I'm actually really grateful I was able to do this at all).  I hope you find this contribution helpful.
-David

**Instructions for using Edit Config for adding Widgets**
1. Go to Config -> Update Configuration -> Edit Config
![Screen Shot 2022-07-11 at 8 10 00 PM](https://user-images.githubusercontent.com/92620199/178379171-aeefd940-d603-45ac-b590-14bffe06eac9.png)
![Screen Shot 2022-07-11 at 8 10 42 PM](https://user-images.githubusercontent.com/92620199/178379247-4b211cf5-f51a-4fad-94f4-e9b6aead51cf.png)

2. Expand the section you want to add the widget(s) to ("Testing Widgets" in this case); if you want to use a new section add that via the GUI first
![Screen Shot 2022-07-11 at 7 26 25 PM](https://user-images.githubusercontent.com/92620199/178378370-ba73a63b-7552-4915-8f65-017a3ca0ec4a.png)

3. Click on the Edit button for the section’s name and select Insert -> Array; You will see 2 new rows were added: “field” in a box with “[0]” next to it and “(empty array)”
![Screen Shot 2022-07-11 at 7 26 52 PM](https://user-images.githubusercontent.com/92620199/178378424-a4c96691-b887-448e-8981-a2c276de6436.png)

4. Type “widgets” into the newly created field
![Screen Shot 2022-07-11 at 7 28 16 PM](https://user-images.githubusercontent.com/92620199/178377924-59555583-82f7-417c-85da-35e1e5791da0.png)

5. Click on the Edit button for the newly created “empty object” and select Append -> Object; You will see 1 new row added with "(empty array)".
![Screen Shot 2022-07-11 at 7 27 49 PM](https://user-images.githubusercontent.com/92620199/178378081-c23f08dc-d4b9-40c1-87d8-c46755ee8cef.png)

6. Click on the Edit button for the newly created “empty object” and select Append -> Auto; See new row with field box and value box.
![Screen Shot 2022-07-11 at 7 32 17 PM](https://user-images.githubusercontent.com/92620199/178377313-f292df04-04be-4f77-8780-6141d3eb3ffa.png)

7. Enter “type” into the field box and enter the name of the widget into the value box.
![Screen Shot 2022-07-11 at 7 32 35 PM](https://user-images.githubusercontent.com/92620199/178377213-1b8f4ab5-1a11-4873-a53c-13be69909846.png)

8. Click on the Edit button for the “type” row and select either Append -> Object (for value fields required by the widget) or Append -> Auto (for boolean fields required by the widget) 

9. Don't forget to hit the Save Changes button and then the Save to Disk button (or whatever save mechanism you prefer) so you don't lose your work.

- You can add additional widgets by repeating steps 5-8.

- You can add an additional widget block by repeating steps 3-8.

-  You can drag the widget and items (buttons) around into water order you want.

- You can drag the name row above the widget row for neatness
![Screen Shot 2022-07-11 at 7 36 26 PM](https://user-images.githubusercontent.com/92620199/178377141-8abeccc6-e498-4ed1-b659-d58ca201100a.png)

**Comment by lissy93** [comment]
Thanks @poke0 for the nice guide :)
I'm gonna transfer it to the discussions section....

Just to add, a much easier way of [editing config](https://github.com/Lissy93/dashy/blob/master/docs/configuring.md#there-are-three-ways-to-edit-the-config), is to work with your config file directly.
Using a tool like [Code-Server](https://github.com/coder/code-server), makes this even easier.

---

Or, using the UI JSON editor, you can also switch to code mode, and quickly copy/pasta stuff in.
YAML and JSON follow the same structure, if you're not so familiar you can just use [this tool](https://jsonformatter.org/yaml-to-json) to convert between the two.

<img src="https://user-images.githubusercontent.com/1862727/178494848-e76eb8f4-d33e-41ef-93d4-7331dd24339c.png" width="500" />

**Comment by poke0** [comment]
I considered Code-Server but it looked too much like a project that I had to figure out before I could play with the widget.  I did try directly editing in code view but I failed miserably.  The only challenge with the Edit Config view was understanding the required relationships between the Autos, Arrays, Objects, and Strings.  Once I got that to work once it was easy sailing, but getting that to work the first time was very frustrating.  Thanks to the pointer to YAML <--> JSON, if I knew about that previously it probably would have changed my approach.  :)
